Crouching beside vehicle holes

joveeater opened the pull request:
#### Summary

SUMMARY: Bugfixes "Crouching next to a vehicle hole limits visibility appropriately"

#### Purpose of change

Fixes the bug where you could see through vehicle holes between transparent parts even when crouching should have rendered them opaque and changes coverage for vehicle parts so that protrusions don't provide cover and so don't affect crouching vision (same as aisles don't currently). It also incidentally fixes an almost unnoticeable issue that until you actually look at a z level, it will still be using uninitialized all solid transparency. 

#### Describe the solution

This hauls out the old crouching visibility system, which was a separate transparency cache per z level for the player. Every turn the level that you're looking at would be copied from the main cache and maybe have the 9 tiles around the player altered. It wasn't used for anything but crouching and making the player's tile transparent. 

This uses a single cache that describes vision adjustments for the 8 tiles around the player. As with the previous logic, tiles can be left alone or made solid but now they can also be hidden. Then, when we're doing the player's vision, these changes are applied to the normal transparency and blocked caches in place, taking care to store any changed values so the caches can be restored afterwards. 

In some ways it's backwards but tiles that are both hidden and made solid are just made solid. This is to avoid the case where crouching anywhere in a rotated car means you can't see the tiles on one diagonal. Technically they're between two opaque things and should be hidden but it looks weird. 

It also looks kinda weird when having a wing mirror lets you see a tile outside the car when you're crouch driving, hence the protrusion change. Crouching behind a wing mirror just seems wrong in general. I took a quick look through and while with some protrusions like the external tank or seed drill it's plausible you could hide behind them, most of them it's not and I wanted to keep the rule simple. 

#### Describe alternatives you've considered

We could go the other way and also produce a player only diagonal blocks cache, or leave the transparency cache in place and only handle diagonal blocks by changing and restoring. If there's something planned that would really use it like x-ray specs that's probably better but as is it's a ~140K per turn memcpy and ~2M of storage for just crouching.

Coolthulhu commented:
Seems that it conflicts with #1743

I'll check the other PR now because it's older. If you think this one should go first, I'll go with this one first in the next run.

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -1277,6 +1304,76 @@ castLightAll<float, float, shrapnel_calc, shrapnel_check,
+        diagonal_blocks( &blocked_cache )[MAPSIZE_X][MAPSIZE_Y],
+        float ( &vision_restore_cache )[9], bool ( &blocked_restore_cache )[8] )
+{
+    const map &here = get_map();
```
Avoid using global functions and variables where possible. It makes refactoring more complex.

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -1277,6 +1304,76 @@ castLightAll<float, float, shrapnel_calc, shrapnel_check,
+        diagonal_blocks( &blocked_cache )[MAPSIZE_X][MAPSIZE_Y],
+        float ( &vision_restore_cache )[9], bool ( &blocked_restore_cache )[8] )
+{
+    const map &here = get_map();
```
Same here. While at the moment, vision functions aren't called on maps that aren't "the global map", passing a `map &` doesn't make the significantly less readable, but it does make it more resilient to a class of bugs.

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -164,34 +164,62 @@ bool map::build_transparency_cache( const int zlev )
+    bool dirty = false;
 
-    const tripoint &p = g->u.pos();
+    if( g->u.movement_mode_is( CMM_CROUCH ) ) {
```
You should pass the player from above, as an argument.
Right now, we have plenty of hacks that use globals, but this is bad practice.

joveeater commented:
They are done. I've moved the vision apply/restore methods into map itself. I was reluctant to add anything more to it but it only makes sense it's alongside the other vision cache building functions. I've moved the global read from building the vision cache up into rebuilding the map caches. It might want moving further up but I wouldn't treat the idea of it as a blanket way to avoid the global accessors. It only makes sense here because building a vision cache for another character is very reasonable. The problems with global state come from how wide reaching it is, not where it comes from. Passing an object around everywhere a global might be used has all the same problems. 

Games almost always end up with all state being effectively global. It's never impossible to imagine that someone will want to make a thing of any type (like an item, mutation, enemy) that affects any other type of thing. Especially good, interconnected game design is often at odds with good separation of concerns. Some game engines will go the pass locals route and they end up with Game being passed on a lot of calls and little else is different.

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -642,8 +670,7 @@ map::apparent_light_info map::apparent_light_helper( const level_cache &map_cach
     auto is_opaque = [&map_cache]( const point & p ) {
-        return map_cache.transparency_cache[p.x][p.y] <= LIGHT_TRANSPARENCY_SOLID &&
-               map_cache.vision_transparency_cache[p.x][p.y] <= LIGHT_TRANSPARENCY_SOLID;
+        return map_cache.transparency_cache[p.x][p.y] <= LIGHT_TRANSPARENCY_SOLID && g->u.pos().xy() != p;
```
`g->u` here should be very explicit and passed from above.

Coolthulhu replied:
You should pass the point from above. If not the point, the `Character` instead.

Why do you compare the exact position here? What would happen if there was no position comparison here?

Coolthulhu replied:
Still a problem.

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -1302,6 +1398,14 @@ void map::build_seen_cache( const tripoint &origin, const int target_z )
+    bool blocked_restore_cache[8] = {false};
+
+    if( origin.z == target_z ) {
+        apply_vision_transparency_cache( g->u.pos(), target_z, vision_restore_cache,
```
`g->u`

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -1340,6 +1444,11 @@ void map::build_seen_cache( const tripoint &origin, const int target_z )
     }
 
+    if( origin.z == target_z ) {
+        restore_vision_transparency_cache( g->u.pos(), target_z, vision_restore_cache,
```
`g->u`

joveeater commented:
Oh yeah, nearly forgot about this, my bad.

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -164,34 +164,62 @@ bool map::build_transparency_cache( const int zlev )
 }
 
-bool map::build_vision_transparency_cache( const int zlev )
+bool map::build_vision_transparency_cache( Character &player )
```
Is the `player` actually modified here? Couldn't it be `const`?

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -1302,6 +1398,14 @@ void map::build_seen_cache( const tripoint &origin, const int target_z )
+    bool blocked_restore_cache[8] = {false};
+
+    if( origin.z == target_z ) {
+        apply_vision_transparency_cache( get_avatar().pos(), target_z, vision_restore_cache,
```
If you're using only `pos`, it should be `get_player_character`.

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -1277,6 +1304,75 @@ castLightAll<float, float, shrapnel_calc, shrapnel_check,
+void map::apply_vision_transparency_cache( const tripoint &center, int target_z,
+        float ( &vision_restore_cache )[9], bool ( &blocked_restore_cache )[8] )
+{
+    auto &map_cache = get_cache( target_z );
```
Would be better as actual type here.
